### PR TITLE
(PUP-9189) Refactor the state transition code in Windows::Service

### PR DESCRIPTION
**Pull request by ekinanp**
Previously, the transition_service_state helper used by the
Windows::Service.start and Windows::Service.stop methods had
a verbose interface that made it a bit more confusing to
understand what was going on than it needed to be. For example,
it was not clear what the difference between the preceding_transition
and preceding_state parameters is from the parameter names alone.

This commit refactors the transition_service_state helper to more
consistently model the Windows service state transition diagram. It
also extends the helper to support the possibility that the service
may have already transitioned to the final state, whereby it will
return early (without performing the transition), or that it is is in
the middle of a pending transition to the final state, whereby it
will wait for the pending transition to finish, and then return early
(without performing the transition). This lets Windows::Service.start
and Windows::Service.stop noop if the service is already starting/started
or stopping/stopped, which is consistent with e.g. systemd.

### CODE REVIEW

File: spec/unit/util/windows/service_spec.rb [hunk 1]
```diff
@@ -90,262 +98,370 @@ def expect_successful_config_query_and_return(query_return)
...
-    context "when the service control manager cannot be opened" do
-      let(:scm) { FFI::Pointer::NULL_HANDLE }
-      it "raises a puppet error" do
-        expect{ subject.start(mock_service_name) }.to raise_error(Puppet::Error)
+      subject.stubs(:ControlService).with(service, signal, anything).returns(FFI::WIN32_FALSE)
+
+      expect { subject.send(action, mock_service_name) }.to raise_error do |error|
+        expect(error).to be_a(Puppet::Error)
+
+        expect(error.message).to match(subject::SERVICE_CONTROL_SIGNALS[signal].to_s)
```
Comment (McdonaldSeanp) [inline spec/unit/util/windows/service_spec.rb]: TBH I don't like matching on any part of the error message. I understand where you're trying to go with this (i.e. trying to remove false positives where something else errors and the test passes) but I don't think it's worth making the test brittle.

I've seen a couple new tests already break, especially with i18n enabled windows machines (which we test in both unit and acceptance).

I believe it's acceptable risk to just unit test that an error is raised.

File: lib/puppet/util/windows/service.rb [hunk 2]
```diff
@@ -266,34 +293,41 @@ def exists?(service_name)
...
-          FFI::MemoryPointer.new(SERVICE_STATUS.size) do |status_ptr|
-            status = SERVICE_STATUS.new(status_ptr)
-            if ControlService(service, SERVICE_CONTROL_STOP, status) == FFI::WIN32_FALSE
-              raise Puppet::Util::Windows::Error.new(_("Failed to send stop control to service, current state is %{current_state}. Failed with") % { current_state: status[:dwCurrentState].to_s })
-            end
-          end
-        end
+      Puppet.debug _("Stopping the %{service_name} service") % { service_name: service_name }
+
+      valid_initial_states = SERVICE_STATES.keys - [SERVICE_STOPPED]
```
Comment (McdonaldSeanp) [inline lib/puppet/util/windows/service.rb]: Is there a specific reason SERVICE_STOPPED isn't a valid initial state?

(and same question for SERVICE_RUNNING in .start)

File: spec/unit/util/windows/service_spec.rb (hunk 1 above)
Comment (ekinanp) [reply spec/unit/util/windows/service_spec.rb]: Yeah that's fine. I'll go ahead and update the PR to remove the error message matching

File: lib/puppet/util/windows/service.rb (hunk 2 above)
Comment (ekinanp) [reply lib/puppet/util/windows/service.rb]: I was thinking it would make it easier to reason about `transition_service_state` by treating the final state separately from the initial state, i.e. so it's something like:
# If the service is in the final state, no further work needs to be done
# Validate the service's initial state
# If it's already transitioning to the final state, wait then return
# Else, perform the transition and wait

whereas adding the final state as an initial state would make it something like:
# Validate the service's initial state
# If the service is in the final state, no further work needs to be done
# If it's already transitioning to the final state, wait then return
# Else, perform the transition and wait

(i.e. validate should move up, otherwise there would be no point in adding the final state in the set of initial states since we will never reach the validation code).

In other words, an initial state in the first set of semantics is a state that _can_ transition to the final state. Having the final state be an initial state there is kind of weird

Comment (McdonaldSeanp) [reply lib/puppet/util/windows/service.rb]: Alrighty, that makes sense.

**Comment by ekinanp** [comment]
Updated the PR to remove the error message matching. There's still some small mistakes I'll need to update before merging this (found out while working on PUP-6702), so adding a Do Not Merge label for now.

**Comment by ekinanp** [comment]
Ok, PR updated. Fixes were all in the unit tests. TLDR of the changes were:

* Removed the service control signal shared example and instead just inlined those tests. The shared example didn't give a lot of flexibility if the block to transition_state_helper does other stuff besides calling ControlService (which it will when implementing `#resume` part of PUP-6702). I did not want to add that flexibility either because the tests there are short enough that inlining them is probably the easiest option.

* Slightly refactored the main `service action that peforms a state transition` shared example to specify the initial_state and a lambda for mocking the state transition itself (i.e. lambda for mocking the block that the transition_service_state helper receives). This makes it more flexible and easier to understand.

Here's my [WIP](https://github.com/ekinanp/puppet/commit/PUP-6702) commit for PUP-6702 that I will open up a PR for once this is merged. The unit tests for `#resume` now pass with these changes.

This is ready for merge (again).

File: lib/puppet/util/windows/service.rb [hunk 3]
```diff
@@ -596,75 +680,109 @@ def query_config(service)
...
-      #   'false' if the service never reached one of the acceptable_states
-      def wait_for_state_transition(service, acceptable_states)
+      # @param [Integer] initial_state the state that the service is transitioning from.
+      # @param [Integer] final_state the state that the service is transitioning to
+      def wait_on_state_transition(service, initial_state, final_state)
+        # Get the pending state for this transition
+        if final_state != SERVICE_RUNNING
+          pending_state = FINAL_STATES.key(final_state)
+        elsif initial_state == SERVICE_STOPPED
+          # final_state == SERVICE_RUNNING
```
[the remark this answers is not in the document]
Comment (caseywilliams) [reply lib/puppet/util/windows/service.rb]: Ok, I agree that that's a nice thing to have in a comment, but when looking it over I just assumed they were unused - I didn't understand that that was information about the state of the service

File: lib/puppet/util/windows/service.rb [hunk 4]
```diff
@@ -596,75 +680,109 @@ def query_config(service)
...
-          # 2. if progress has been made, increment elapsed_time and set last_checkpoint
-          # 3. sleep, then loop again if there was progress.
-          time_to_wait = wait_hint_to_wait_time(status[:dwWaitHint])
+
+          # Check if any progress has been made since our last sleep
+          # using the dwCheckPoint. If no progress has been made then
+          # check if we've timed out, and raise an error if so
           if status[:dwCheckPoint] > last_checkpoint
             elapsed_time = 0
+            last_checkpoint = status[:dwCheckPoint]
```
Comment (speedofdark) [inline lib/puppet/util/windows/service.rb]: Since we set elapsed_time to 0 in line 750 and dont modify it before line 768, do we need both?

File: lib/puppet/util/windows/service.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/puppet/util/windows/service.rb]: I can clarify them a bit more. Will push an update in a second

File: lib/puppet/util/windows/service.rb (hunk 4 above)
Comment (ekinanp) [reply lib/puppet/util/windows/service.rb]: Yes. Line 750 sets it to the scope outside of the begin block. Otherwise, the elapsed_time variable will only be scoped to the begin end block which means it cannot be referenced in latter iterations of the loop.

File: lib/puppet/util/windows/service.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/puppet/util/windows/service.rb]: Updated. Are they clearer now?

[the remark this answers is not in the document]
Comment (caseywilliams) [reply lib/puppet/util/windows/service.rb]: That makes sense, thanks

File: lib/puppet/util/windows/service.rb (hunk 4 above)
Comment (speedofdark) [reply lib/puppet/util/windows/service.rb]: Ya. I get that. I was wondering why it needed to be (re)set inside the block when there is nothing that changes it in-between the two referenced locations.

Comment (speedofdark) [reply lib/puppet/util/windows/service.rb]: Aha. I get it now. Nvm.
